### PR TITLE
Add FacetSplitPC

### PR DESCRIPTION
**Pull request by pbrubeck**
A Python PC that splits a FunctionSpace into cell interior DOFs and facet DOFs.

### CODE REVIEW

File: firedrake/preconditioners/facet_split.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    perm = numpy.empty((V.dof_count, ), dtype=PETSc.IntType)
+    perm.fill(-1)
+
+    offset = 0
+    for Wsub in W:
+        W_indices = None
+        W_map, nel = glonum_fun(Wsub.cell_node_map(), bsize=bsize)
+        rdofs = restricted_dofs(Wsub.finat_element, V.finat_element)
+
+        for e in range(nel):
```
Comment (dham) [inline firedrake/preconditioners/facet_split.py:24]: Element loop in Python code is usually a problem. Are we really sure this won't be expensive for large meshes? Should it not be in cython?

File: firedrake/preconditioners/facet_split.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        v.project(w_expr, solver_parameters={
+                  "mat_type": "matfree",
+                  "ksp_type": "cg",
+                  "ksp_atol": 0,
+                  "ksp_rtol": rtol,
+                  "pc_type": "jacobi", })
+    return numpy.rint((end-start-1)*v.dat.data_ro.reshape((-1,))+start).astype(PETSc.IntType)
+
+
+class FacetSplitPC(PCBase):
```
Comment (dham) [inline firedrake/preconditioners/facet_split.py:60]: Please provide documentation and tests.

File: firedrake/preconditioners/facet_split.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                sub_cells = cell.sub_cells()
+                degree = max(e.degree())
+                variant = e.variant()
+                Qc_elt = FiniteElement("Q", sub_cells[0], degree, variant=variant)
+                Qd_elt = FiniteElement("RTCE", sub_cells[0], degree, variant=variant)
+                Id_elt = FiniteElement("DG", sub_cells[1], degree - 1, variant=variant)
+                Ic_elt = FiniteElement("CG", sub_cells[1], degree, variant=variant)
+                return EnrichedElement(HCurl(TensorProductElement(FacetElement(Qc_elt), Id_elt, cell=cell)),
+                                       HCurl(TensorProductElement(Qd_elt, FacetElement(Ic_elt), cell=cell)),
+                                       HCurl(TensorProductElement(FacetElement(Qd_elt), InteriorElement(Ic_elt), cell=cell)))
```
Comment (wence-) [inline firedrake/preconditioners/facet_split.py:128]: Does "just" making a `RestrictedElement(element, restriction="facet")` not work?

Comment (pbrubeck) [reply firedrake/preconditioners/facet_split.py:128]: This does not work, see here #2713

Comment (wence-) [reply firedrake/preconditioners/facet_split.py:128]: Does the addition of finat/finat#105 now fix this?

Comment (pbrubeck) [reply firedrake/preconditioners/facet_split.py]: yes

File: firedrake/preconditioners/facet_split.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+                 v.dat(op2.WRITE, pmap),
+                 w.dat[0](op2.READ, W[0].cell_node_map()),
+                 w.dat[1](op2.READ, W[1].cell_node_map()))
+
+    own = V.dof_dset.set.size * bsize
+    perm = V.dof_dset.lgmap.apply(perm, result=perm)
+    return perm[:own]
+
+
+def get_permutation_project(V, W):
```
Comment (pefarrell) [inline firedrake/preconditioners/facet_split.py:272]: Can you add a comment here?

**Review by pefarrell — CHANGES_REQUESTED** [review]

**Review by pefarrell — DISMISSED** [review]

File: firedrake/preconditioners/facet_split.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    _prefix = "facet_"
+
+    _permutation_cache = {}
+
+    def get_permutation(self, V, W):
+        from mpi4py import MPI
+        key = (V, W)
+        if key not in self._permutation_cache:
+            indices = get_permutation_map(V, W)
+            if V.comm.allreduce(numpy.all(indices[:-1] <= indices[1:]), MPI.PROD):
```
Comment (JDBetteridge) [inline firedrake/preconditioners/facet_split.py:33]: `V._comm`

File: firedrake/preconditioners/facet_split.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+                self._permutation_cache[key] = None
+            else:
+                self._permutation_cache[key] = indices
+        return self._permutation_cache[key]
+
+    def initialize(self, pc):
+        from ufl import RestrictedElement, MixedElement, TensorElement, VectorElement
+        from firedrake import FunctionSpace, TestFunctions, TrialFunctions
+        from firedrake.assemble import allocate_matrix, TwoFormAssembler
+        from functools import partial
```
Comment (JDBetteridge) [inline firedrake/preconditioners/facet_split.py:43]: Move (all) imports to the top of the file

File: firedrake/preconditioners/facet_split.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        assert W.dim() == V.dim(), "Dimensions of the original and decomposed spaces do not match"
+
+        mixed_operator = a(sum(TestFunctions(W)), sum(TrialFunctions(W)), coefficients={})
+        mixed_bcs = tuple(bc.reconstruct(V=W[-1], g=0) for bc in bcs)
+
+        self.perm = None
+        self.iperm = None
+        indices = self.get_permutation(V, W)
+        if indices is not None:
+            self.perm = PETSc.IS().createGeneral(indices, comm=V.comm)
```
Comment (JDBetteridge) [inline firedrake/preconditioners/facet_split.py:85]: `V._comm`

**Review by JDBetteridge — CHANGES_REQUESTED** [review]
imports need moving to the top, some comms need to be updated

File: firedrake/preconditioners/facet_split.py (hunk 6 above)
Comment (pbrubeck) [reply firedrake/preconditioners/facet_split.py:43]: I get a circular import error if I do that for `firedrake.assemble`.

Comment (JDBetteridge) [reply firedrake/preconditioners/facet_split.py]: Ah, leave the assemble where it is then!

File: firedrake/preconditioners/facet_split.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    own = V.dof_dset.layout_vec.sizes[0]
+    perm = perm.reshape((-1, ))
+    perm = V.dof_dset.lgmap.apply(perm, result=perm)
+    return perm[:own]
+
+
+def get_permutation_project(V, W):
+    """ Alternative projection-based method to obtain DOF permutation
+    """
+    from firedrake import Function
```
Comment (dham) [inline firedrake/preconditioners/facet_split.py:271]: Please to be removing dead code.

**Review by dham — CHANGES_REQUESTED** [review]

**Review by dham — APPROVED** [review]

**Pull request merged**